### PR TITLE
Use setpoints averaged in a zone time step to calculate "hours of setpoint not met" with onoff thermostat

### PR DESCRIPTION
**Pull request by lgu1234**
Pull request overview
---------------------
ZoneThermostatSetPointLo and ZoneThermostatSetPointHi are used to calculate "hours of setpoint not met" at every zone time step. When onoff thermostat is applied, ZoneThermostatSetPointLo and ZoneThermostatSetPointHi are calculated at every system time step. The setpoint averaged in a zone time step is used.

This fix addresses the last item in #6546.
### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by nrel-bot-2** [comment]
@lgu1234 @lgentile it has been 15 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@lgu1234 @lgentile it has been 14 days since this pull request was last updated.

File: tst/EnergyPlus/unit/ThermalComfort.unit.cc [hunk 1]
```diff
@@ -890,3 +891,61 @@ TEST_F(EnergyPlusFixture, ThermalComfort_CalcThermalComfortAdaptiveASH55Test)
...
+    EXPECT_EQ(0, ThermalComfortSetPoint(1).notMetHeating);
+    EXPECT_EQ(0, ThermalComfortSetPoint(1).notMetHeatingOccupied);
+    EXPECT_EQ(0, ThermalComfortSetPoint(1).notMetCooling);
+    EXPECT_EQ(0, ThermalComfortSetPoint(1).notMetCoolingOccupied);
+    EXPECT_EQ(TimeStepZone, ThermalComfortSetPoint(1).totalNotMetHeating);
+    EXPECT_EQ(TimeStepZone, ThermalComfortSetPoint(1).totalNotMetHeatingOccupied);
+    EXPECT_EQ(TimeStepZone, ThermalComfortSetPoint(1).totalNotMetCooling);
+    EXPECT_EQ(TimeStepZone, ThermalComfortSetPoint(1).totalNotMetCoolingOccupied);
+
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ThermalComfort.unit.cc:951]: I had expected this unit test to have multiple zones based on the pull request, but I guess the unit test is actually testing the right thing here.  This PR isn't so much about the averaging calculation of multiple zones, but rather what E+ does after it's calculated.  This unit test is good.

File: src/EnergyPlus/DataHeatBalFanSys.cc [hunk 2]
```diff
@@ -212,6 +212,8 @@ namespace DataHeatBalFanSys {
     Array1D<Real64> AdapComfortCoolingSetPoint;
     Array1D<Real64> ZoneThermostatSetPointHi;
     Array1D<Real64> ZoneThermostatSetPointLo;
+    Array1D<Real64> ZoneThermostatSetPointHiAver;
+    Array1D<Real64> ZoneThermostatSetPointLoAver;
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalFanSys.cc:216]: More arrays that shouldn't be standalone arrays, but should instead be member variables on the zone structure.  In the future, **please** start setting the precedent for moving these types of things.

File: src/EnergyPlus/DataHeatBalFanSys.cc [hunk 3]
```diff
@@ -332,6 +334,8 @@ namespace DataHeatBalFanSys {
         AdapComfortCoolingSetPoint.deallocate();
         ZoneThermostatSetPointHi.deallocate();
         ZoneThermostatSetPointLo.deallocate();
+        ZoneThermostatSetPointHiAver.deallocate();
+        ZoneThermostatSetPointLoAver.deallocate();
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalFanSys.cc:338]: Added to `clear_state`, good.

File: src/EnergyPlus/ThermalComfort.cc [hunk 4]
```diff
@@ -2453,7 +2457,11 @@ namespace ThermalComfort {
                 if (AirModel(iZone).AirModelType != RoomAirModel_Mixing) {
                     deltaT = TempTstatAir(iZone) - ZoneThermostatSetPointLo(iZone);
                 } else {
-                    deltaT = ZTAV(iZone) - ZoneThermostatSetPointLo(iZone);
+                    if (NumOnOffCtrZone > 0) {
+                        deltaT = ZTAV(iZone) - ZoneThermostatSetPointLoAver(iZone);
+                    } else {
+                        deltaT = ZTAV(iZone) - ZoneThermostatSetPointLo(iZone);
```
Comment (Myoldmopar) [inline src/EnergyPlus/ThermalComfort.cc:2463]: So can you explain why other thermostat types don't also benefit from using the average zone temp?

**Review by Myoldmopar — COMMENTED** [review]
I think this code is ready to go.  Some minor comments.  I would like to hear your response to my comment about other thermostat types before merging.

File: src/EnergyPlus/DataHeatBalFanSys.cc (hunk 2 above)
Comment (lgu1234) [reply src/EnergyPlus/DataHeatBalFanSys.cc:216]: @Myoldmopar I am OK with your suggestions. I ma going to post a new issue to consolidate these variables under Zone struct, such as 

ZoneThermostatSetPointHi;
ZoneThermostatSetPointLo;
ZoneThermostatSetPointHiAver;
ZoneThermostatSetPointLoAver

File: src/EnergyPlus/ThermalComfort.cc (hunk 4 above)
Comment (lgu1234) [reply src/EnergyPlus/ThermalComfort.cc:2463]: The variables with other thermostat types are based on each zone time step, while ZTAV is also zone time step based. Therefore, no need to make any changes. When on/off thermostat is used, thermal setpoint is calculated at each system time step. That is why the average values at every zone time step should be used.

**Comment by Myoldmopar** [comment]
Thanks @lgu1234 that makes sense.  I'll merge.

**Pull request merged**